libwasmvm/src/error/go.rs

```diff
@@ -15,77 +15,185 @@ use crate::memory::UnmanagedVector;
     User = 5,
+    /// An error type that should never be created by us. It only serves as a fallback for the i32 to GoError conversion.
+    Other = -1,
 }
 
-impl From<i32> for GoResult {
+impl From<i32> for GoError {
     fn from(n: i32) -> Self {
-        use GoResult::*;
         // This conversion treats any number that is not otherwise an expected value as `GoError::Other`
         match n {
-            0 => Ok,
-            1 => Panic,
-            2 => BadArgument,
-            3 => OutOfGas,
-            5 => User,
-            _ => Other,
+            0 => GoError::None,
+            1 => GoError::Panic,
+            2 => GoError::BadArgument,
+            3 => GoError::OutOfGas,
+            4 => GoError::CannotSerialize,
+            5 => GoError::User,
+            _ => GoError::Other,
         }
     }
 }
 
-impl fmt::Display for GoResult {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self {
-            GoResult::Ok => write!(f, "Ok"),
-            GoResult::Panic => write!(f, "Panic"),
-            GoResult::BadArgument => write!(f, "BadArgument"),
-            GoResult::OutOfGas => write!(f, "OutOfGas"),
-            GoResult::Other => write!(f, "Other Error"),
-            GoResult::User => write!(f, "User Error"),
-        }
-    }
-}
-
-impl GoResult {
-    /// This converts a GoResult to a `Result<(), BackendError>`, using a fallback error message for some cases.
-    /// If it is GoResult::User the error message will be returned to the contract.
+impl GoError {
+    /// This converts a GoError to a `Result<(), BackendError>`, using a fallback error message for some cases.
+    /// If it is GoError::User the error message will be returned to the contract.
     /// Otherwise, the returned error will trigger a trap in the VM and abort contract execution immediately.
     ///
-    /// Safety: this reads data from an externally provided buffer and assumes valid utf-8 encoding
-    /// Only call if you trust the code that provides `error_msg` to be correct.
-    pub unsafe fn into_ffi_result<F>(
+    /// This reads data from an externally provided `UnmanagedVector` and assumes UFT-8 encoding. To protect
+    /// against invalid UTF-8 data, a lossy conversion to string is used. The data is limited to 8KB in order
+    /// to protect against long externally generated error messages.
+    ///
+    /// The `error_msg` is always consumed here and must not be used afterwards.
+    pub unsafe fn into_result<F>(
         self,
         error_msg: UnmanagedVector,
-        default: F,
+        default_error_msg: F,
     ) -> Result<(), BackendError>
     where
-        F: Fn() -> String,
+        F: FnOnce() -> String,
     {
-        let read_error_msg = || -> String {
-            match error_msg.consume() {
-                Some(data) => String::from_utf8_lossy(&data).into(),
-                None => default(),
+        const MAX_ERROR_LEN: usize = 8 * 1024;
```

Comment:
Nice control check here.
It if it too long, it may cut in the middle of a UTF-8 code point and render the string invalid. But I guess that would be just the same as explictly erroring on too long errors.